src/transaction/log_replication_atomic.cpp

```diff
@@ -116,7 +116,25 @@ namespace cublog
 		    thread_entry, m_redo_lsa);
 	    break;
 	  case LOG_ABORT:
-	    assert (!m_atomic_helper.is_part_of_atomic_replication (header.trid));
+	    // TODO: there are 2 identified sources for aborted transactions:
+	    //  *1* transactions aborted by the client
+	    //  *2* unilaterally aborted transactions by the engine (for whatever reason,
+	    //    eg. log recovery)
+	    // in both these cases, there can remain suspended atomic sequences
+	    // which need to be applied; however at abort it is already too late
+	    // to identify this state of a transaction (eg: at least for *2* the
+	    // recovery state could be identified by processing the log compensate records;
+	    //
+	    // for *2*, the issue http://jira.cubrid.org/browse/LETS-572 has been added;
+	    //
+	    // for now, a naive aproach of asserting and forcibly eliminating the atomic sequence
+	    assert (!m_atomic_helper.is_part_of_atomic_replication (header.trid)
+		    || m_atomic_helper.all_log_entries_are_control (header.trid));
+	    if (m_atomic_helper.is_part_of_atomic_replication (header.trid)
+		&& m_atomic_helper.all_log_entries_are_control (header.trid))
+	      {
+		m_atomic_helper.forcibly_remove_sequence (header.trid);
```

Comment:
I think there could be other cases in which an atomic sequence doesn't complete properly without LOG_ABORT, so the sequence has to be cleaned up. As far as I know, there are two cases in which a transaction can be rolled back to some point and continues. It is (1) SAVEPOINT and (2) sysop abort (end with LOG_SYSOP_END_ABORT).
(1) With SAVEPOINT, a user or a developer put a save point and can roll back to that point and continue the transaction. See xtran_server_savepoint().
(2) After a sysop starts, the sysop can abort anytime for some reason. Then, the transaction is rolled back to when the sysop starts and appends LOG_SYSOP_END_ABORT regardless of its intended ending. In this case, the transaction usually are going to be aborted, but it's not guaranteed. The transaction's destiny is just decided by the operation's error handling.

I don't think the SAVEPOINT can cut in an atomic system operation since sysop is used for an internal operation and SAVEPOINT is used by a user (developer or end-user). However, the second can be the case. Even if it always ends up aborting when atomic sysops are aborted for now, I think we should deal with it because there it is allowed architecturally in my opinion.

Reply:
I took the liberty of pinging back from http://jira.cubrid.org/browse/LETS-572 to this comment.
This will be taken up when that is handled.
Or do you think is more urgent and should be handled asap?

Reply:
It's not about "urgent", but I meant these cases can happen and they should be handled, even if they don't happen for now. If they will be handled before release and are tracked correctly, it's OK.

Reply:
Agreed, I overlooked, but there is already an issue - http://jira.cubrid.org/browse/LETS-541 - within the same epic that says exactly this. That some of the `SYSOP_END`'s are not handled.

Reply:
Oh, I see.